frontend/fs.go

```diff
@@ -0,0 +1,211 @@
+package frontend
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"path"
+	"sync"
+
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/tonistiigi/fsutil/types"
+
+	"github.com/tonistiigi/fsutil"
+
+	"github.com/moby/buildkit/client/llb"
+)
+
+var _ fs.DirEntry = &stateRefDirEntry{}
+var _ fs.ReadDirFS = &StateRefFS{}
+var _ io.ReaderAt = &stateRefFile{}
+
+type StateRefFS struct {
+	s         llb.State
+	ctx       context.Context
+	opts      []llb.ConstraintsOpt
+	client    gwclient.Client
+	clientRes *gwclient.Result
+	o         sync.Once
+	initErr   error
+}
+
+func NewStateRefFS(s llb.State, ctx context.Context, client gwclient.Client) *StateRefFS {
```

Comment:
Maybe `FromRef(context.Context, gwclient.Reference)`

With that something like`FromState(ctx llb.State, client gwclient.Client)` can be a convenience wrapper/

With `FromState` I think we should do the solve up front rather than lazily.

Reply:
This is done, let us know if this is what you were thinking.